Mathlib/RingTheory/MvPolynomial/NewtonIdentities.lean

```diff
@@ -0,0 +1,327 @@
+    simp_rw [← pow_add, add_comm]
+    have h3 : card t.fst ≥ 1
+    · have h4 : t.fst.Nonempty
+      · use t.snd
+        apply h1
+      exact lt_iff_add_one_le.mp (card_pos.mpr h4)
+    rw [← tsub_tsub_assoc h.right.left h3,
+      ← neg_neg ((-1 : MvPolynomial σ R) ^ (card t.fst - 1)), h2 (card t.fst - 1),
+      Nat.sub_add_cancel]
+    simp
+    exact h3
+  · simp_rw [card_insert_of_not_mem h1, prod_insert h1, mul_comm, mul_assoc (∏ a in t.fst, X a),
+      ← mul_add]
+    nth_rewrite 2 [← pow_one (X t.snd)]
+    have h3 : card t.fst + 1 ≤ k
+    · have ht1 := h.right.right
+      contrapose! ht1
+      exact And.intro (le_antisymm h.right.left (le_of_lt_succ ht1)) h1
+    simp_rw [← pow_add, ← Nat.add_sub_assoc h3, add_comm, add_tsub_add_eq_tsub_right]
+    rw [← neg_neg ((-1 : MvPolynomial σ R) ^ (card t.fst)), h2]
+    simp
+
+theorem weight_zero_for_fixed_by_T (t : Finset σ × σ) (h : t ∈ pairs σ k)
+    (h1 : weight σ R k t ≠ 0) : T_map_restr σ t h ≠ t := by
+  by_contra h2
+  have h3 := weight_compose_T σ R t h
+  rw [h2, ← two_mul, _root_.mul_eq_zero] at h3
+  cases' h3 with hl hr
+  case inl => exact two_ne_zero hl
+  case inr => exact h1 hr
+
+theorem weight_sum (k : ℕ) : ∑ t in pairs σ k, weight σ R k t = 0 := by
+  exact sum_involution (T_map_restr σ) (weight_compose_T σ R) (weight_zero_for_fixed_by_T σ R)
+    (T_map_pair σ) (T_map_invol σ)
+
+open Classical in
+theorem sum_equiv_k (k : ℕ) (f : Finset σ × σ → MvPolynomial σ R) :
+    (∑ t in filter (fun t ↦ card t.fst = k) (pairs σ k), f t) =
+    ∑ A in powersetLen k univ, (∑ j in A, f (A, j)) := by
+  apply sum_finset_product
+  simp_all
+  intro p b
+  simp_rw [pairs, mem_filter, PairsPred]
+  simp_all
+  apply Iff.intro
+  · intro hpl
+    simp_all
+    exact mem_powerset_len_univ_iff.mpr hpl.2
+  · intro hpr
+    simp_all
+    have cardpk := mem_powerset_len_univ_iff.mp hpr.1
+    exact And.intro (le_of_eq cardpk) cardpk
+
+open Classical in
+theorem sum_equiv_i_lt_k (k i : ℕ) (hi : i ∈ range k) (f : Finset σ × σ → MvPolynomial σ R) :
+    (∑ t in filter (fun t ↦ card t.fst = i) (pairs σ k), f t) =
+    ∑ A in powersetLen i univ, (∑ j, f (A, j)) := by
+  apply sum_finset_product
+  simp_all
+  intro p b
+  apply Iff.intro
+  · intro hpl
+    exact mem_powerset_len_univ_iff.mpr hpl.2
+  · intro hpr
+    simp_rw [pairs, mem_filter, PairsPred]
+    simp_all
+    apply And.intro
+    · apply And.intro
+      · simp_rw [mem_powerset_len_univ_iff.mp hpr, le_iff_lt_or_eq]
+        left
+        exact hi
+      · intro cardpk
+        rw [← cardpk, mem_powerset_len_univ_iff.mp hpr] at hi
+        exact ((lt_irrefl _) hi).elim
+    · exact mem_powerset_len_univ_iff.mp hpr
+
+theorem sum_equiv_lt_k (k : ℕ) (f : Finset σ × σ → MvPolynomial σ R) :
+    (∑ t in filter (fun t ↦ card t.fst < k) (pairs σ k), f t) =
+    ∑ i in range k, ∑ A in powersetLen i univ, (∑ j, f (A, j)) := by
+    have equiv_i (i : ℕ) (hi : i ∈ range k) := sum_equiv_i_lt_k σ R k i hi f
```

Comment:
The proof should be indented two spaces rather than four

Reply:
Good catch! Fixed